File: C++/16.解释器模式/InterpreterPatternDemo.cpp
/**
 * @file InterpreterPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-15
 *
 * @copyright Copyright (c) 2019
 *
 */
#include "AndExpression.h"
#include "OrExpression.h"
#include "TerminalExpression.h"

static std::shared_ptr<Expression> GetMaleExpression() {
    std::shared_ptr<Expression> robert = std::make_shared<TerminalExpression>("Robert");
    std::shared_ptr<Expression> john = std::make_shared<TerminalExpression>("John");
    return std::make_shared<OrExpression>(robert, john);
}

static std::shared_ptr<Expression> GetMarriedWomanExpression() {
    std::shared_ptr<Expression> julie = std::make_shared<TerminalExpression>("Julie");
    std::shared_ptr<Expression> married = std::make_shared<TerminalExpression>("Married");
    return std::make_shared<AndExpression>(julie, married);
}
int main(int argc, char const* argv[]) {
    std::shared_ptr<Expression> is_male = GetMaleExpression();
    std::shared_ptr<Expression> is_married_woman = GetMarriedWomanExpression();

    std::cout << "John is male? " << is_male->Interpret("John") << std::endl;
    std::cout << "Julie is a married women? " << is_married_woman->Interpret("Married Julie") << std::endl;

    return 0;
}

File: Objective-C/16.InterpreterPattern/InterpreterPatternDemo.m
//
//  InterpreterPatternDemo.m
//  16.InterpreterPattern
//
//  Created by Layne on 2020/11/6.
//

#import "AndExpression.h"
#import "OrExpression.h"
#import "TerminalExpression.h"

static id<ExpressionProtocol> getMaleExpression() {
    id<ExpressionProtocol> robert = [[TerminalExpression alloc] initWithData:@"Robert"];
    id<ExpressionProtocol> john = [[TerminalExpression alloc] initWithData:@"John"];
    return [[OrExpression alloc] initWithExpr1:robert Expr2:john];
}

static id<ExpressionProtocol> getMarriedWomanExpression() {
    id<ExpressionProtocol> julie = [[TerminalExpression alloc] initWithData:@"Julie"];
    id<ExpressionProtocol> married = [[TerminalExpression alloc] initWithData:@"Married"];
    return [[AndExpression alloc] initWithExpr1:julie Expr2:married];
}

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        id<ExpressionProtocol> isMale = getMaleExpression();
        id<ExpressionProtocol> isMarriedWoman = getMarriedWomanExpression();

        NSLog(@"John is male? %@", [isMale interpret:@"John"] ? @"true" : @"false");
        NSLog(@"Julie is a married women? %@", [isMarriedWoman interpret:@"Married Julie"] ? @"true" : @"false");
    }
    return 0;
}
